flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/slotpool/DeclarativeSlotPoolBridgeResourceDeclarationTest.java

```diff
@@ -0,0 +1,154 @@
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeoutException;
+
+import static org.apache.flink.runtime.jobmaster.slotpool.DeclarativeSlotPoolBridgeTest.createAllocatedSlot;
+import static org.apache.flink.runtime.jobmaster.slotpool.DeclarativeSlotPoolBridgeTest.createDeclarativeSlotPoolBridge;
+import static org.hamcrest.CoreMatchers.is;
+import static org.junit.Assert.assertThat;
+
+/**
+ * Tests for the {@link DeclarativeSlotPoolBridge}.
+ */
+public class DeclarativeSlotPoolBridgeResourceDeclarationTest extends TestLogger {
+
+	private static final JobMasterId jobMasterId = JobMasterId.generate();
+	private final ComponentMainThreadExecutor mainThreadExecutor = ComponentMainThreadExecutorServiceAdapter.forMainThread();
+
+	private RequirementListener requirementListener;
+	private DeclarativeSlotPoolBridge declarativeSlotPoolBridge;
+
+	@Before
+	public void setup() throws Exception {
+		requirementListener = new RequirementListener();
+
+		final TestingDeclarativeSlotPoolBuilder slotPoolBuilder = TestingDeclarativeSlotPool.builder()
+			.setIncreaseResourceRequirementsByConsumer(requirementListener::increaseRequirements)
+			.setDecreaseResourceRequirementsByConsumer(requirementListener::decreaseRequirements)
+			.setReserveFreeSlotFunction((allocationId, resourceProfile) -> createAllocatedSlot(allocationId))
+			.setFreeReservedSlotFunction((allocationID, throwable, aLong) -> ResourceCounter.withResource(ResourceProfile.UNKNOWN, 1))
+			.setReleaseSlotFunction((allocationID, e) -> ResourceCounter.withResource(ResourceProfile.UNKNOWN, 1));
+
+		final DeclarativeSlotPoolBridgeTest.TestingDeclarativeSlotPoolFactory declarativeSlotPoolFactory = new DeclarativeSlotPoolBridgeTest.TestingDeclarativeSlotPoolFactory(slotPoolBuilder);
+		declarativeSlotPoolBridge = createDeclarativeSlotPoolBridge(declarativeSlotPoolFactory);
+		declarativeSlotPoolBridge.start(jobMasterId, "localhost", mainThreadExecutor);
+	}
+
+	@After
+	public void teardown() throws Exception {
+		if (declarativeSlotPoolBridge != null) {
+			declarativeSlotPoolBridge.close();
+		}
+	}
+
+	@Test
+	public void testRequirementsIncreasedOnNewAllocation() {
+		// requesting the allocation of a new slot should increase the requirements
+		declarativeSlotPoolBridge.requestNewAllocatedSlot(new SlotRequestId(), ResourceProfile.UNKNOWN, Time.minutes(5));
+		assertThat(requirementListener.getRequirements().getResourceCount(ResourceProfile.UNKNOWN), is(1));
+	}
+
+	@Test
+	public void testRequirementsDecreasedOnAllocationTimeout() {
+		// requesting the allocation of a new slot increases the requirements
+		final CompletableFuture<PhysicalSlot> allocationFuture = declarativeSlotPoolBridge.requestNewAllocatedSlot(new SlotRequestId(), ResourceProfile.UNKNOWN, Time.minutes(5));
+		// when the allocation fails the requirements should be reduced (it is the users responsibility to retry)
+		allocationFuture.completeExceptionally(new TimeoutException());
```

Comment:
I think the fact that failing this future triggers the cleanup is an implementation detail which can change in the future. I would suggest to use a very small timeout so that the request fails fast.